### PR TITLE
reqrite: better event emitter; update main class

### PR DESCRIPTION
**Pull request by mathe42**
build ontop of #33 and #40

(depends on #40 to be merged then the dfif will be much smaller!)

removes callback hell :+1: 

closes #40 
closes #44

### CODE REVIEW

**Comment by mathe42** [comment]
@oberbeck and another one :D.

**Comment by mathe42** [comment]
@oberbeck what is the differences with your PR and this one (when we ignore index.ts)?

**Review by oberbeck — COMMENTED** [review]
Just one note on generics; I prefer a naming convention where the generics are recognisable as such and declarative.

Ie:
```ts
function<TType, TName>(type: TType, name: TName){}
```

Like a T prefix as a common convention for generics and then the name of what it is representing.

Of course, this is highly opinionated - I just wanted to take the time and share my thoughts.

**TLDR;** looks good to me 👍

File: src/index.ts [hunk 1]
```diff
@@ -80,7 +80,23 @@ export type Auth =
...
+	}
+	& {
+		[
+			K in Exclude<
+				string,
+				"open" | "opened" | "close" | "closed" | "notify"
+			>
+		]: [Result<any>];
+	}
+> {
```
Comment (oberbeck) [inline src/index.ts:99]: @mathe42 you asked for a difference; this is a little redundant on a nit-picky level. The reason for this is the anonymous type as it can't be reused for the Exclude.

This is my approach:
https://github.com/surrealdb/surrealdb.js/pull/44/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R83-R99

So for in your case, what do you think about:

```ts
interface Whatever {
 		open: [];
 		opened: [];
 		close: [];
 		closed: [];
 		notify: [any];
}

export default class Surreal extends Emitter<
 	& Whatever
 	& {
 		[
 			K in Exclude<
 				string,
 				keyof Whatever
 			>
 		]: [Result<any>];
 	}
 > {
 ```

**Comment by mathe42** [comment]
Yes and I hate that convention...

File: src/classes/emitter.ts [hunk 2]
```diff
@@ -1,49 +1,87 @@
...
+	#events: {
+		[K in keyof Events]?: Set<(this: this, ...args: Events[K]) => void>;
+	} = {};
+
+	static once = once;
 
-	on(e: string, func: Listener) {
-		if (typeof this.#events[e] !== "object") {
-			this.#events[e] = [];
+	next<T extends keyof Events>(eventName: T) {
```
[the remark this answers is not in the document]
Comment (mathe42) [reply src/classes/emitter.ts:20]: Naming can be changed. It is basicly a pattern but not a well known one. In other projects I added a helper `nextEvent` or similar. Got the idea beacause of the code of @Choooks22. (See index.ts how it effects the code)

File: src/classes/emitter.ts [hunk 3]
```diff
@@ -1,49 +1,87 @@
...
+
+	on<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	) {
+		if (!this.#events[eventName]) {
+			this.#events[eventName] = new Set();
 		}
-		this.#events[e].push(func);
+		this.#events[eventName]!.add(listener);
```
[the remark this answers is not in the document]
Comment (mathe42) [reply src/classes/emitter.ts:31]: As we left the if all types are as before. But we know it is defined now.

File: src/index.ts (hunk 1 above)
Comment (mathe42) [reply src/index.ts:99]: Yes agree I will change that.

File: src/classes/emitter.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mathe42) [reply src/classes/emitter.ts:20]: Im not well known in ReactiveX/Angular (used Angular 4 a while ago but swiched to VueJS). But I'm happy to rename that function. What is a good name?

`nextEvent`, `nextPromise`,...

**Comment by mathe42** [comment]
@tobiemh 

Ready for review / merge: 

This is a 2 part PR:

1. rewrite of the eventEmitter with better TypeScript types. (based on #33)
2. rewrite of the index.ts to use the `nextEvent` method of the eventEmitter (nextEvent based on #33)
